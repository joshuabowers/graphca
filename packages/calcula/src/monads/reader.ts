export type Reader<E, T> = (env: E) => T

export const unit = <E, T>(t: T): Reader<E, T> => _ => t

export const bind = <E, A>(r: Reader<E, A>) => 
  <B>(fn: (a: A) => Reader<E, B>): Reader<E, B> =>
    (e: E) => fn(r(e))(e)

export const ask = <E>(env: E) => env

export const local = <E, T>(fn: (env: E) => E) =>
  (c: Reader<E, T>) =>
    (e: E) => c(fn(e))
