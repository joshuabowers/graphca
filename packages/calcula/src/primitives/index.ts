import { isPrimitive, PrimitiveNode } from '../closures/primitive'
import { Real, real, isReal, EulerMascheroni } from './real'
import { 
  Complex, complex, isComplex, ComplexInfinity, isComplexInfinity 
} from './complex'
import { Boolean, boolean, isBoolean } from './boolean'
import { Nil, nil, isNil } from './nil'
import { NaN, nan, isNaN } from './nan'

export type Primitive = Real | Complex | Boolean | Nil | NaN
export { Real, Complex, Boolean, Nil, NaN }

export type PrimitiveFns = 
| typeof real | typeof complex | typeof boolean

export { real, complex, boolean, nil, nan }
export { isReal, isComplex, isBoolean, isNil, isNaN, isComplexInfinity }
export { isPrimitive, PrimitiveNode }
export { EulerMascheroni, ComplexInfinity }
