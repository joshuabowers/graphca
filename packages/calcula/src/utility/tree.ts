import { Writer } from "../monads/writer"

export enum Clades {
  primitive = 'Primitive',  // Constants: 10, 4 - 5i, true
  unary = 'Unary',          // Functions and Ops: not, abs, sin
  binary = 'Binary',        // Functions and Ops: and, log, polygamma
  variadic = 'Variadic'     // Variables: x, OhGodWhy
}

export enum Genera {
  arithmetic = 'Arithmetic',
  connective = 'Connective',
  inequalities = 'Inequalities',
  trigonometric = 'Trigonometric',
  arcus = 'Arcus',
  hyperbolic = 'Hyperbolic',
  areaHyperbolic = 'AreaHyperbolic',
  combinatorics = 'Combinatorics'
}

export enum Species {
  // Primitives
  real = 'Real',
  complex = 'Complex',
  boolean = 'Boolean',
  nil = 'Nil',
  nan = 'NaN',

  // Arithmetic
  add = 'Addition',
  multiply = 'Multiplication',
  raise = 'Exponentiation',

  // Functions
  abs = 'Absolute',

  // :: Inequalities
  equals = 'Equality',
  notEquals = 'Inequality',
  lessThan = 'LessThan',
  greaterThan = 'GreaterThan',
  lessThanEquals = 'LessThanEquals',
  greaterThanEquals = 'GreaterThanEquals',

  // :: Connectives
  not = 'Complement',
  and = 'Conjunction',
  or = 'Disjunction',
  xor = 'ExclusiveDisjunction',
  implies = 'Implication',
  nand = 'AlternativeDenial',
  nor = 'JointDenial',
  xnor = 'Biconditional',
  converse = 'ConverseImplication',

  // :: Trigonometric
  sin = 'Sine',
  cos = 'Cosine',
  tan = 'Tangent',
  csc = 'Cosecant',
  sec = 'Secant',
  cot = 'Cotangent',

  // Variables
  variable = 'Variable'
}

export type TreeNode = {
  readonly clade: Clades,
  readonly genus?: Genera,
  readonly species: Species
}

export type DerivedNode<T extends ((...args: any[]) => any)> = 
  ReturnType<T> extends Writer<infer U> ? U : never

export type TreeNodeGuardFn<T extends TreeNode> = 
  (value: Writer<TreeNode>) => value is Writer<T>

// values should be Writer<TreeNode>, narrowed to Writer<T>
export const isClade = <T extends TreeNode>(clade: Clades) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.clade === clade

export const isGenus = <T extends TreeNode>(genus: Genera) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.genus === genus

export const isSpecies = <T extends TreeNode>(species: Species) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.species === species

export const notAny = (...args: Species[]) => {
  const exclude = new Set(args)
  return <T extends TreeNode>(t: Writer<T>) => 
    t.value && !exclude.has(t.value.species)
}

export const any = (...args: Species[]) => {
  const include = new Set(args)
  return <T extends TreeNode>(t: Writer<T>) => 
    t.value && include.has(t.value.species)
}
