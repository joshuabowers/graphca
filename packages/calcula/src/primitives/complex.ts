import { Species } from "../utility/tree"
import { Complex, primitive } from "../closures/primitive"
import { isNumberTuple } from "../utility/valuePredicates"
export { Complex }

export const [complex, isComplex, $complex] = 
  primitive<[number, number], {a: number, b: number}, Complex>(
  isNumberTuple,
  ([a, b]) => ({a, b}),
  Species.complex
)(
  r => [r.value, 0],
  c => [c.a, c.b],
  b => [b.value ? 1 : 0, 0]
)()

export const ComplexInfinity = complex([Infinity, NaN])
