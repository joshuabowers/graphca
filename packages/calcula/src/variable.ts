import { Writer, unit } from './monads/writer'
import { TreeNode, Clades, Species, isSpecies } from './utility/tree'
import { isNil, nil } from './primitives/nil'
import { rule, process, resolve } from './utility/rule'

export type Variable = TreeNode & {
  readonly clade: Clades.variadic,
  readonly species: Species.variable,
  readonly name: string,
  readonly value: Writer<TreeNode>
}

export const isVariable = isSpecies<Variable>(Species.variable)

export const variable = (
  name: string, value: Writer<TreeNode> = nil
): Writer<Variable> => 
  ({
    value: {
      clade: Clades.variadic,
      species: Species.variable,
      name, value
    },
    log: [{
      input: process`${name}`, 
      rewrite: isNil(value) ? resolve`${name}` : process`${value}`,
      action: isNil(value) 
        ? 'referenced variable' : 'referenced variable with value'
    }] 
  })

export type Scope = Map<string, Writer<Variable>>

export const isScope = (v: unknown): v is Scope =>
  typeof v === 'object' && v !== null
  && ['get', 'set', 'has', 'delete'].every(m => m in v)

type Entries = Iterable<readonly [string, Writer<Variable>]>

export const scope = (entries: Entries = []): Scope => 
  new Map<string, Writer<Variable>>(entries)

export const assign = (name: string, value: Writer<TreeNode>, scope?: Scope) => {
  const v = variable(name, value)
  if(scope){
    if(isNil(value)){ scope.delete(name) }
    else { scope.set(name, v) }
  }
  return v
}
