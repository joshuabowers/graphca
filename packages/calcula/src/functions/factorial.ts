import { unit } from "../monads/writer";
import { Species, Notation } from "../utility/tree";
import { ComplexInfinity } from '../primitives/complex';
import { real, complex, boolean, isReal, isComplex } from '../primitives'
import { Unary, unary, when, UnaryNodeMetaTuple, unaryPostfixRule } from "../closures/unary";
import { add, multiply } from "../arithmetic";
import { gamma } from './gamma';
import { isValue } from "../utility/deepEquals";
import { rule } from "../utility/rule";
import { Unicode } from "../Unicode";

export type Factorial = Unary<Species.factorial>

export const factorialRule = unaryPostfixRule('!')

const isNegativeInteger = (t: number) => t < 0 && Number.isInteger(t)

export const [factorial, isFactorial, $factorial] = unary<Factorial>(
  '!', Notation.postfix, Species.factorial
)(
  r => multiply(unit(r), factorial(add(unit(r), real(-1)))), 
  c => multiply(unit(c), factorial(add(unit(c), complex([-1, 0])))), 
  _b => boolean(true),
)(
  when(
    t => isReal(t) && isNegativeInteger(t.value.value), 
    [ComplexInfinity, rule`${ComplexInfinity}`, 'singularity'])
  ,
  when(
    t => isComplex(t) && isNegativeInteger(t.value.a) && t.value.b === 0, 
    [ComplexInfinity, rule`${ComplexInfinity}`, 'singularity']
  ),
  when(
    t => isReal(t) && !Number.isInteger(t.value.value),
    r => [
      gamma(add(unit(r), real(1))), 
      rule`${Unicode.gamma}(${r} + ${real(1)})`, 
      'calculated factorial via gamma'
    ]
  ),
  when(
    t => isComplex(t) && (t.value.b !== 0 || !Number.isInteger(t.value.a)),
    c => [
      gamma(add(unit(c), complex([1, 0]))), 
      rule`${Unicode.gamma}(${c} + ${complex([1, 0])})`,
      'calculated factorial via gamma'
    ]
  ),
  when(isValue(real(0)), [real(1), rule`${real(1)}`, 'degenerate case']),
  when(isValue(complex([0, 0])), [complex([1, 0]), rule`${complex([1, 0])}`, 'degenerate case'])
) as UnaryNodeMetaTuple<Factorial, void> // ffs
