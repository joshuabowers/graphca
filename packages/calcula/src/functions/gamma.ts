import { Writer, unit } from "../monads/writer"
import { Multi, multi, method } from "@arrows/multimethod"
import { TreeNode, Species } from "../utility/tree"
import { Real, Complex, real, boolean, isReal, isComplex } from "../primitives"
import { Unary, unary, when } from "../closures/unary"
import { 
  add, subtract, multiply, divide, negate, raise, sqrt 
} from "../arithmetic"
import { sin } from "./trigonometric"
import { factorial } from "./factorial"

const isPIN = (n: number) => n > 0 && n <= 15 && Number.isInteger(n)

const isPositiveInteger = (e: Writer<TreeNode>) =>
  (isReal(e) && isPIN(e.value.value))
  || (isComplex(e) && e.value.b === 0 && isPIN(e.value.a))


export type Predicate<T extends TreeNode> = (t: T) => boolean

export type ConstantPredicate = Multi 
  & Predicate<Real>
  & Predicate<Complex>  

const isSmall: ConstantPredicate = multi(
  method(isReal, (r: Writer<Real>) => r.value.value < 0.5),
  method(isComplex, (c: Writer<Complex>) => c.value.a < 0.5),
  method(false)
)

const lanczos = {
  p: [
    676.5203681218851,
    -1259.1392167224028,
    771.32342877765313,
    -176.61502916214059,
    12.507343278686905,
    -0.13857109526572012,
    9.9843695780195716e-6,
    1.5056327351493116e-7
  ]
}

const pi = real(Math.PI), sqrtTwicePi = sqrt(real(2 * Math.PI))
const gammaReflection = (e: Writer<TreeNode>): Writer<TreeNode> =>
  divide(
    pi,
    multiply(sin(multiply(e, pi)), gamma(subtract(real(1), e)))
  )

const calculateGamma = (input: Writer<Real|Complex>): Writer<TreeNode> => {
  const one = real(1)
  const z = subtract(input, one)
  const x = lanczos.p.reduce(
    (s, v, i) => add(s, divide(real(v), add(z, add(real(i), one)))),
    real(0.99999999999980993) as Writer<TreeNode>
  )
  const t = subtract(add(z, real(lanczos.p.length)), real(0.5))
  return multiply(
    sqrtTwicePi,
    multiply(
      raise(t, add(z, real(0.5))),
      multiply(raise(real(Math.E), negate(t)), x)
    )
  )
}

export type Gamma = Unary<Species.gamma>

export const [gamma, isGamma, $gamma] = unary<Gamma>(Species.gamma)(
  r => [calculateGamma(unit(r)) as Writer<Real>, 'computed real gamma'],
  c => [calculateGamma(unit(c)) as Writer<Complex>, 'computed complex gamma'],
  b => [boolean(calculateGamma(real(b.value ? 1 : 0)) as Writer<Real>), 'computed boolean gamma']
)(
  when(
    isPositiveInteger, 
    t => [factorial(subtract(unit(t), real(1))), 'computing gamma via factorial']
  ),
  when(
    t => (isReal(t) && t.value.value < 0.5)
      || (isComplex(t) && t.value.a < 0.5),
    t => [gammaReflection(unit(t)), 'gamma reflection for small value']
  )
)
