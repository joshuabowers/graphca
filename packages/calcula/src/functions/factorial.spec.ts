import { unit } from '../monads/writer';
import { 
  expectToEqualWithSnapshot, expectCloseTo
} from '../utility/expectations'
import { Clades, Species } from '../utility/tree';
import { ComplexInfinity } from '../primitives/complex';
import { real, complex } from '../primitives';
import { variable } from '../variable';
import { factorial, $factorial } from "./factorial";
import { gamma } from './gamma'

describe('$factorial', () => {
  it('generates a Factorial for a TreeNode input', () => {
    expect(
      $factorial(unit(variable('x').value))[0]
    ).toEqual({
      clade: Clades.unary, genus: undefined, species: Species.factorial,
      expression: unit(variable('x').value)
    })
  })
})

describe('factorial', () => {
  it('returns 1 for an input of 0', () => {
    expectToEqualWithSnapshot(
      factorial(real(0)),
      real(1)
    )
  })

  it('returns the factorial for positive integers', () => {
    expectToEqualWithSnapshot(
      factorial(real(2)),
      real(2)
    )
  })

  it('returns complex 1 for an input of complex 0', () => {
    expectToEqualWithSnapshot(
      factorial(complex([0, 0])),
      complex([1, 0])
    )
  })

  it('returns the factorial for a integer complex', () => {
    expectToEqualWithSnapshot(
      factorial(complex([2, 0])),
      complex([2, 0])
    )
  })

  it('returns complex infinity for non-positive integer reals', () => {
    expectToEqualWithSnapshot(
      factorial(real(-5)),
      ComplexInfinity
    )
  })

  it('returns a shifted gamma for non-integer reals', () => {
    expectCloseTo(factorial(real(5.5)), gamma(real(6.5)), 10)
  })

  it('returns a shifted gamma for non-integer complex numbers', () => {
    expectCloseTo(factorial(complex([1, 1])), gamma(complex([2, 1])), 10)
  })

  it('returns a Factorial node for unbound variables', () => {
    expectToEqualWithSnapshot(
      factorial(variable('x')),
      $factorial(variable('x'))[0]
    )
  })
})
