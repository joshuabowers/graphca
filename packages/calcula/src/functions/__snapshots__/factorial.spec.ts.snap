// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factorial returns 1 for an input of 0 1`] = `
[
  [
    "(0)!",
    "identified factorial",
  ],
  [
    "([0])!",
    "processing argument",
  ],
  [
    "0",
    "created real",
  ],
  [
    "({0})!",
    "processed argument",
  ],
  [
    "(0)!",
    "degenerate case",
  ],
  [
    "1",
    "created real",
  ],
]
`;

exports[`factorial returns a Factorial node for unbound variables 1`] = `
[
  [
    "(x)!",
    "identified factorial",
  ],
  [
    "([x])!",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x})!",
    "processed argument",
  ],
  [
    "(x)!",
    "created factorial",
  ],
]
`;

exports[`factorial returns complex 1 for an input of complex 0 1`] = `
[
  [
    "(0+0𝒊)!",
    "identified factorial",
  ],
  [
    "([0+0𝒊])!",
    "processing argument",
  ],
  [
    "0+0𝒊",
    "created complex",
  ],
  [
    "({0+0𝒊})!",
    "processed argument",
  ],
  [
    "(0+0𝒊)!",
    "degenerate case",
  ],
  [
    "1+0𝒊",
    "created complex",
  ],
]
`;

exports[`factorial returns complex infinity for non-positive integer reals 1`] = `
[
  [
    "(-5)!",
    "identified factorial",
  ],
  [
    "([-5])!",
    "processing argument",
  ],
  [
    "-5",
    "created real",
  ],
  [
    "({-5})!",
    "processed argument",
  ],
  [
    "(-5)!",
    "singularity",
  ],
  [
    "∞̄",
    "created complex",
  ],
]
`;

exports[`factorial returns the factorial for a integer complex 1`] = `
[
  [
    "(2+0𝒊)!",
    "identified factorial",
  ],
  [
    "([2+0𝒊])!",
    "processing argument",
  ],
  [
    "2+0𝒊",
    "created complex",
  ],
  [
    "({2+0𝒊})!",
    "processed argument",
  ],
  [
    "(2+0𝒊)!",
    "complex factorial",
  ],
  [
    "(2+0𝒊*((2+0𝒊+-1+0𝒊))!)",
    "identified multiplication",
  ],
  [
    "([2+0𝒊]*((2+0𝒊+-1+0𝒊))!)",
    "processing left operand",
  ],
  [
    "2+0𝒊",
    "created complex",
  ],
  [
    "({2+0𝒊}*[((2+0𝒊+-1+0𝒊))!])",
    "processed left operand; processing right operand",
  ],
  [
    "((2+0𝒊+-1+0𝒊))!",
    "identified factorial",
  ],
  [
    "([(2+0𝒊+-1+0𝒊)])!",
    "processing argument",
  ],
  [
    "(2+0𝒊+-1+0𝒊)",
    "identified addition",
  ],
  [
    "([2+0𝒊]+-1+0𝒊)",
    "processing left operand",
  ],
  [
    "2+0𝒊",
    "created complex",
  ],
  [
    "({2+0𝒊}+[-1+0𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "-1+0𝒊",
    "created complex",
  ],
  [
    "(2+0𝒊+{-1+0𝒊})",
    "processed right operand",
  ],
  [
    "(2+0𝒊+-1+0𝒊)",
    "complex addition",
  ],
  [
    "1+0𝒊",
    "created complex",
  ],
  [
    "({1+0𝒊})!",
    "processed argument",
  ],
  [
    "(1+0𝒊)!",
    "complex factorial",
  ],
  [
    "(1+0𝒊*((1+0𝒊+-1+0𝒊))!)",
    "identified multiplication",
  ],
  [
    "([1+0𝒊]*((1+0𝒊+-1+0𝒊))!)",
    "processing left operand",
  ],
  [
    "1+0𝒊",
    "created complex",
  ],
  [
    "({1+0𝒊}*[((1+0𝒊+-1+0𝒊))!])",
    "processed left operand; processing right operand",
  ],
  [
    "((1+0𝒊+-1+0𝒊))!",
    "identified factorial",
  ],
  [
    "([(1+0𝒊+-1+0𝒊)])!",
    "processing argument",
  ],
  [
    "(1+0𝒊+-1+0𝒊)",
    "identified addition",
  ],
  [
    "([1+0𝒊]+-1+0𝒊)",
    "processing left operand",
  ],
  [
    "1+0𝒊",
    "created complex",
  ],
  [
    "({1+0𝒊}+[-1+0𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "-1+0𝒊",
    "created complex",
  ],
  [
    "(1+0𝒊+{-1+0𝒊})",
    "processed right operand",
  ],
  [
    "(1+0𝒊+-1+0𝒊)",
    "complex addition",
  ],
  [
    "0+0𝒊",
    "created complex",
  ],
  [
    "({0+0𝒊})!",
    "processed argument",
  ],
  [
    "(0+0𝒊)!",
    "degenerate case",
  ],
  [
    "1+0𝒊",
    "created complex",
  ],
  [
    "(1+0𝒊*{1+0𝒊})",
    "processed right operand",
  ],
  [
    "(1+0𝒊*1+0𝒊)",
    "complex multiplication",
  ],
  [
    "1+0𝒊",
    "created complex",
  ],
  [
    "(2+0𝒊*{1+0𝒊})",
    "processed right operand",
  ],
  [
    "(2+0𝒊*1+0𝒊)",
    "complex multiplication",
  ],
  [
    "2+0𝒊",
    "created complex",
  ],
]
`;

exports[`factorial returns the factorial for positive integers 1`] = `
[
  [
    "(2)!",
    "identified factorial",
  ],
  [
    "([2])!",
    "processing argument",
  ],
  [
    "2",
    "created real",
  ],
  [
    "({2})!",
    "processed argument",
  ],
  [
    "(2)!",
    "real factorial",
  ],
  [
    "(2*((2+-1))!)",
    "identified multiplication",
  ],
  [
    "([2]*((2+-1))!)",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "({2}*[((2+-1))!])",
    "processed left operand; processing right operand",
  ],
  [
    "((2+-1))!",
    "identified factorial",
  ],
  [
    "([(2+-1)])!",
    "processing argument",
  ],
  [
    "(2+-1)",
    "identified addition",
  ],
  [
    "([2]+-1)",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "({2}+[-1])",
    "processed left operand; processing right operand",
  ],
  [
    "-1",
    "created real",
  ],
  [
    "(2+{-1})",
    "processed right operand",
  ],
  [
    "(2+-1)",
    "real addition",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1})!",
    "processed argument",
  ],
  [
    "(1)!",
    "real factorial",
  ],
  [
    "(1*((1+-1))!)",
    "identified multiplication",
  ],
  [
    "([1]*((1+-1))!)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}*[((1+-1))!])",
    "processed left operand; processing right operand",
  ],
  [
    "((1+-1))!",
    "identified factorial",
  ],
  [
    "([(1+-1)])!",
    "processing argument",
  ],
  [
    "(1+-1)",
    "identified addition",
  ],
  [
    "([1]+-1)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}+[-1])",
    "processed left operand; processing right operand",
  ],
  [
    "-1",
    "created real",
  ],
  [
    "(1+{-1})",
    "processed right operand",
  ],
  [
    "(1+-1)",
    "real addition",
  ],
  [
    "0",
    "created real",
  ],
  [
    "({0})!",
    "processed argument",
  ],
  [
    "(0)!",
    "degenerate case",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(1*{1})",
    "processed right operand",
  ],
  [
    "(1*1)",
    "multiplicative identity",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(2*{1})",
    "processed right operand",
  ],
  [
    "(2*1)",
    "real multiplication",
  ],
  [
    "2",
    "created real",
  ],
]
`;
